### ISSUE TITLE
ConnectionFactory.connect property 'connection.timeout' is insufficient for connection timeout

### ISSUE THREAD

**Issue by cljohnso** [Terracotta-OSS/terracotta-core#232]
Specifying `connection.timeout` on `ConnectionFactory.connect` is used as the timeout value for the
`DistributedObjectClient.waitForConnection` timeout.  This works well enough but `DistributedObjectClient.waitForConnection` is synchronized and competes with `DistributedObjectClient.openChannel` (also synchronized) running in a "Connection Establisher" thread.  The 'openChannel' method ultimately reaches `ClientMessageTransport.waitForAynAck` which uses a timeout value from a statically obtained `TCPropertiesConsts.TC_TRANSPORT_HANDSHAKE_TIMEOUT` property.  If this value is larger than the  `connection.timeout`, the `TCPropertiesConsts.TC_TRANSPORT_HANDSHAKE_TIMEOUT` may prevail.  In addition, the call stack to `DistributedObjectClient.waitForConnection` includes `DistributedObjectClientFactory.create` which, when a connection attempt times out, calls `DistributedObjectClient.shutdown`.  This ultimately invokes `DistributedObjectClient.shutdownResources` which is subject to yet another timeout -- the value of the `TCPropertiesConsts.L1_SHUTDOWN_THREADGROUP_GRACETIME` property.

Bottom line, the `ConnectionFactory.connect` method does not return until potentially each of these timeouts is met.

For predictable operation, each of these connection-related timeouts should be driven from the `connection.timeout`, if any.

**Comment by myronkscott** [comment]
Most of this will be fixed by better synchronization. 

The issue with shutdown i believe must be addressed by ensuring efficient cleanup so that grace time doesn't come into play.

**Comment by cljohnso** [comment]
Lest someone read your comment and get the wrong idea, this isn't something observed with shutdown per-se -- all of this occurs while attempting to recognize that a connection attempt, based on `connection.timeout`, has timed out.  The shutdown is entirely internal.
